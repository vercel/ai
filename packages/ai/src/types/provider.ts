import { EmbeddingModel } from './embedding-model';
import { LanguageModel } from './language-model';
import { ImageModel } from './image-model';

/**
 * Provider for language, text embedding, and image models.
 */
export type Provider = {
  /**
  Returns the language model with the given id.
  The model id is then passed to the provider function to get the model.

  @param {string} id - The id of the model to return.

  @returns {LanguageModel} The language model associated with the id

  @throws {NoSuchModelError} If no such model exists.
     */
  languageModel(modelId: string): LanguageModel;

  /**
  Returns the text embedding model with the given id.
  The model id is then passed to the provider function to get the model.

  @param {string} id - The id of the model to return.

  @returns {LanguageModel} The language model associated with the id

  @throws {NoSuchModelError} If no such model exists.
     */
  textEmbeddingModel(modelId: string): EmbeddingModel<string>;

  /**
  Returns the image model with the given id.
  The model id is then passed to the provider function to get the model.

  @param {string} id - The id of the model to return.

  @returns {ImageModel} The image model associated with the id
  */
  imageModel(modelId: string): ImageModel;
};
