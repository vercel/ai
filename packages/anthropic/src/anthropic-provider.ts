import {
  LanguageModelV1,
  NoSuchModelError,
  ProviderV1,
} from '@ai-sdk/provider';
import {
  FetchFunction,
  loadApiKey,
  withoutTrailingSlash,
} from '@ai-sdk/provider-utils';
import { AnthropicMessagesLanguageModel } from './anthropic-messages-language-model';
import {
  AnthropicMessagesModelId,
  AnthropicMessagesSettings,
} from './anthropic-messages-settings';
import { anthropicTools } from './anthropic-tools';

export interface AnthropicProvider extends ProviderV1 {
  /**
Creates a model for text generation.
*/
  (
    modelId: AnthropicMessagesModelId,
    settings?: AnthropicMessagesSettings,
  ): LanguageModelV1;

  /**
Creates a model for text generation.
*/
  languageModel(
    modelId: AnthropicMessagesModelId,
    settings?: AnthropicMessagesSettings,
  ): LanguageModelV1;

  /**
@deprecated Use `.languageModel()` instead.
*/
  chat(
    modelId: AnthropicMessagesModelId,
    settings?: AnthropicMessagesSettings,
  ): LanguageModelV1;

  /**
@deprecated Use `.languageModel()` instead.
   */
  messages(
    modelId: AnthropicMessagesModelId,
    settings?: AnthropicMessagesSettings,
  ): LanguageModelV1;

  /**
Anthropic-specific computer use tool.
   */
  tools: typeof anthropicTools;
}

export interface AnthropicProviderSettings {
  /**
Use a different URL prefix for API calls, e.g. to use proxy servers.
The default prefix is `https://api.anthropic.com/v1`.
   */
  baseURL?: string;

  /**
API key that is being send using the `x-api-key` header.
It defaults to the `ANTHROPIC_API_KEY` environment variable.
   */
  apiKey?: string;

  /**
Custom headers to include in the requests.
     */
  headers?: Record<string, string>;

  /**
Custom fetch implementation. You can use it as a middleware to intercept requests,
or to provide a custom fetch implementation for e.g. testing.
    */
  fetch?: FetchFunction;

  generateId?: () => string;
}

/**
Create an Anthropic provider instance.
 */
export function createAnthropic(
  options: AnthropicProviderSettings = {},
): AnthropicProvider {
  const baseURL =
    withoutTrailingSlash(options.baseURL) ?? 'https://api.anthropic.com/v1';

  const getHeaders = () => ({
    'anthropic-version': '2023-06-01',
    'x-api-key': loadApiKey({
      apiKey: options.apiKey,
      environmentVariableName: 'ANTHROPIC_API_KEY',
      description: 'Anthropic',
    }),
    ...options.headers,
  });

  const createChatModel = (
    modelId: AnthropicMessagesModelId,
    settings: AnthropicMessagesSettings = {},
  ) =>
    new AnthropicMessagesLanguageModel(modelId, settings, {
      provider: 'anthropic.messages',
      baseURL,
      headers: getHeaders,
      fetch: options.fetch,
      supportsImageUrls: true,
    });

  const provider = function (
    modelId: AnthropicMessagesModelId,
    settings?: AnthropicMessagesSettings,
  ) {
    if (new.target) {
      throw new Error(
        'The Anthropic model function cannot be called with the new keyword.',
      );
    }

    return createChatModel(modelId, settings);
  };

  provider.languageModel = createChatModel;
  provider.chat = createChatModel;
  provider.messages = createChatModel;
  provider.textEmbeddingModel = (modelId: string) => {
    throw new NoSuchModelError({ modelId, modelType: 'textEmbeddingModel' });
  };

  provider.tools = anthropicTools;

  return provider;
}

/**
Default Anthropic provider instance.
 */
export const anthropic = createAnthropic();
