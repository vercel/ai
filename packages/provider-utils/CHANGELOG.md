# @ai-sdk/provider-utils

## 2.2.0

### Minor Changes

- 5bc638d: AI SDK 4.2

### Patch Changes

- Updated dependencies [5bc638d]
  - @ai-sdk/provider@1.1.0

## 2.1.15

### Patch Changes

- d0c4659: feat (provider-utils): parseProviderOptions function

## 2.1.14

### Patch Changes

- Updated dependencies [0bd5bc6]
  - @ai-sdk/provider@1.0.12

## 2.1.13

### Patch Changes

- Updated dependencies [2e1101a]
  - @ai-sdk/provider@1.0.11

## 2.1.12

### Patch Changes

- 1531959: feat (provider-utils): add readable-stream to unified test server

## 2.1.11

### Patch Changes

- Updated dependencies [e1d3d42]
  - @ai-sdk/provider@1.0.10

## 2.1.10

### Patch Changes

- Updated dependencies [ddf9740]
  - @ai-sdk/provider@1.0.9

## 2.1.9

### Patch Changes

- Updated dependencies [2761f06]
  - @ai-sdk/provider@1.0.8

## 2.1.8

### Patch Changes

- 2e898b4: chore (ai): move mockId test helper into provider utils

## 2.1.7

### Patch Changes

- 3ff4ef8: feat (provider-utils): export removeUndefinedEntries for working with e.g. headers

## 2.1.6

### Patch Changes

- Updated dependencies [d89c3b9]
  - @ai-sdk/provider@1.0.7

## 2.1.5

### Patch Changes

- 3a602ca: chore (core): rename CoreTool to Tool

## 2.1.4

### Patch Changes

- 066206e: feat (provider-utils): move delay to provider-utils from ai

## 2.1.3

### Patch Changes

- 39e5c1f: feat (provider-utils): add getFromApi and response handlers for binary responses and status-code errors

## 2.1.2

### Patch Changes

- ed012d2: feat (provider): add metadata extraction mechanism to openai-compatible providers
- Updated dependencies [3a58a2e]
  - @ai-sdk/provider@1.0.6

## 2.1.1

### Patch Changes

- e7a9ec9: feat (provider-utils): include raw value in json parse results
- Updated dependencies [0a699f1]
  - @ai-sdk/provider@1.0.5

## 2.1.0

### Minor Changes

- 62ba5ad: release: AI SDK 4.1

## 2.0.8

### Patch Changes

- 00114c5: feat: expose IDGenerator and createIdGenerator

## 2.0.7

### Patch Changes

- 90fb95a: chore (provider-utils): switch to unified test server
- e6dfef4: feat (provider/fireworks): Support add'l image models.
- 6636db6: feat (provider-utils): add unified test server

## 2.0.6

### Patch Changes

- 19a2ce7: feat (provider/fireworks): Add image model support.
- 6337688: feat: change image generation errors to warnings
- Updated dependencies [19a2ce7]
- Updated dependencies [6337688]
  - @ai-sdk/provider@1.0.4

## 2.0.5

### Patch Changes

- 5ed5e45: chore (config): Use ts-library.json tsconfig for no-UI libs.
- Updated dependencies [5ed5e45]
  - @ai-sdk/provider@1.0.3

## 2.0.4

### Patch Changes

- Updated dependencies [09a9cab]
  - @ai-sdk/provider@1.0.2

## 2.0.3

### Patch Changes

- 0984f0b: feat (provider-utils): Add resolvable type and utility routine.

## 2.0.2

### Patch Changes

- Updated dependencies [b446ae5]
  - @ai-sdk/provider@1.0.1

## 2.0.1

### Patch Changes

- c3ab5de: fix (provider-utils): downgrade nanoid and secure-json-parse (ESM compatibility)

## 2.0.0

### Major Changes

- b469a7e: chore: remove isXXXError methods
- b1da952: chore (provider-utils): remove convertStreamToArray
- 8426f55: chore (ai):increase id generator default size from 7 to 16.
- db46ce5: chore (provider-utils): remove isParseableJson export

### Patch Changes

- dce4158: chore (dependencies): update eventsource-parser to 3.0.0
- dce4158: chore (dependencies): update nanoid to 5.0.8
- Updated dependencies [b469a7e]
- Updated dependencies [c0ddc24]
  - @ai-sdk/provider@1.0.0

## 2.0.0-canary.3

### Major Changes

- 8426f55: chore (ai):increase id generator default size from 7 to 16.

## 2.0.0-canary.2

### Patch Changes

- dce4158: chore (dependencies): update eventsource-parser to 3.0.0
- dce4158: chore (dependencies): update nanoid to 5.0.8

## 2.0.0-canary.1

### Major Changes

- b1da952: chore (provider-utils): remove convertStreamToArray

## 2.0.0-canary.0

### Major Changes

- b469a7e: chore: remove isXXXError methods
- db46ce5: chore (provider-utils): remove isParseableJson export

### Patch Changes

- Updated dependencies [b469a7e]
- Updated dependencies [c0ddc24]
  - @ai-sdk/provider@1.0.0-canary.0

## 1.0.22

### Patch Changes

- aa98cdb: chore: more flexible dependency versioning
- 7b937c5: feat (provider-utils): improve id generator robustness
- 811a317: feat (ai/core): multi-part tool results (incl. images)
- Updated dependencies [aa98cdb]
- Updated dependencies [1486128]
- Updated dependencies [7b937c5]
- Updated dependencies [3b1b69a]
- Updated dependencies [811a317]
  - @ai-sdk/provider@0.0.26

## 1.0.21

### Patch Changes

- Updated dependencies [b9b0d7b]
  - @ai-sdk/provider@0.0.25

## 1.0.20

### Patch Changes

- Updated dependencies [d595d0d]
  - @ai-sdk/provider@0.0.24

## 1.0.19

### Patch Changes

- 273f696: fix (ai/provider-utils): expose size argument in generateId

## 1.0.18

### Patch Changes

- 03313cd: feat (ai): expose response id, response model, response timestamp in telemetry and api
- Updated dependencies [03313cd]
- Updated dependencies [3be7c1c]
  - @ai-sdk/provider@0.0.23

## 1.0.17

### Patch Changes

- Updated dependencies [26515cb]
  - @ai-sdk/provider@0.0.22

## 1.0.16

### Patch Changes

- 09f895f: feat (ai/core): no-schema output for generateObject / streamObject

## 1.0.15

### Patch Changes

- d67fa9c: feat (provider/amazon-bedrock): add support for session tokens

## 1.0.14

### Patch Changes

- Updated dependencies [f2c025e]
  - @ai-sdk/provider@0.0.21

## 1.0.13

### Patch Changes

- Updated dependencies [6ac355e]
  - @ai-sdk/provider@0.0.20

## 1.0.12

### Patch Changes

- dd712ac: fix: use FetchFunction type to prevent self-reference

## 1.0.11

### Patch Changes

- Updated dependencies [dd4a0f5]
  - @ai-sdk/provider@0.0.19

## 1.0.10

### Patch Changes

- 4bd27a9: chore (ai/provider): refactor type validation
- 845754b: fix (ai/provider): fix atob/btoa execution on cloudflare edge workers
- Updated dependencies [4bd27a9]
  - @ai-sdk/provider@0.0.18

## 1.0.9

### Patch Changes

- Updated dependencies [029af4c]
  - @ai-sdk/provider@0.0.17

## 1.0.8

### Patch Changes

- Updated dependencies [d58517b]
  - @ai-sdk/provider@0.0.16

## 1.0.7

### Patch Changes

- Updated dependencies [96aed25]
  - @ai-sdk/provider@0.0.15

## 1.0.6

### Patch Changes

- 9614584: fix (ai/core): use Symbol.for
- 0762a22: feat (ai/core): support zod transformers in generateObject & streamObject

## 1.0.5

### Patch Changes

- a8d1c9e9: feat (ai/core): parallel image download
- Updated dependencies [a8d1c9e9]
  - @ai-sdk/provider@0.0.14

## 1.0.4

### Patch Changes

- 4f88248f: feat (core): support json schema

## 1.0.3

### Patch Changes

- Updated dependencies [2b9da0f0]
- Updated dependencies [a5b58845]
- Updated dependencies [4aa8deb3]
- Updated dependencies [13b27ec6]
  - @ai-sdk/provider@0.0.13

## 1.0.2

### Patch Changes

- Updated dependencies [b7290943]
  - @ai-sdk/provider@0.0.12

## 1.0.1

### Patch Changes

- d481729f: fix (ai/provider-utils): generalize to Error (DomException not always available)

## 1.0.0

### Major Changes

- 5edc6110: feat (provider-utils): change getRequestHeader() test helper to return Record (breaking change)

### Patch Changes

- 5edc6110: feat (provider-utils): add combineHeaders helper
- Updated dependencies [5edc6110]
  - @ai-sdk/provider@0.0.11

## 0.0.16

### Patch Changes

- 02f6a088: feat (provider-utils): add convertArrayToAsyncIterable test helper

## 0.0.15

### Patch Changes

- 85712895: feat (@ai-sdk/provider-utils): add createJsonStreamResponseHandler
- 85712895: chore (@ai-sdk/provider-utils): move test helper to provider utils

## 0.0.14

### Patch Changes

- 7910ae84: feat (providers): support custom fetch implementations

## 0.0.13

### Patch Changes

- Updated dependencies [102ca22f]
  - @ai-sdk/provider@0.0.10

## 0.0.12

### Patch Changes

- 09295e2e: feat (@ai-sdk/provider-utils): add download helper
- 043a5de2: fix (provider-utils): rename to isParsableJson
- Updated dependencies [09295e2e]
  - @ai-sdk/provider@0.0.9

## 0.0.11

### Patch Changes

- Updated dependencies [f39c0dd2]
  - @ai-sdk/provider@0.0.8

## 0.0.10

### Patch Changes

- Updated dependencies [8e780288]
  - @ai-sdk/provider@0.0.7

## 0.0.9

### Patch Changes

- 6a50ac4: feat (provider-utils): add loadSetting and convertAsyncGeneratorToReadableStream helpers
- Updated dependencies [6a50ac4]
  - @ai-sdk/provider@0.0.6

## 0.0.8

### Patch Changes

- Updated dependencies [0f6bc4e]
  - @ai-sdk/provider@0.0.5

## 0.0.7

### Patch Changes

- Updated dependencies [325ca55]
  - @ai-sdk/provider@0.0.4

## 0.0.6

### Patch Changes

- 276f22b: fix (ai/provider): improve request error handling

## 0.0.5

### Patch Changes

- Updated dependencies [41d5736]
  - @ai-sdk/provider@0.0.3

## 0.0.4

### Patch Changes

- 56ef84a: ai/core: fix abort handling in transformation stream

## 0.0.3

### Patch Changes

- 25f3350: ai/core: add support for getting raw response headers.
- Updated dependencies [d6431ae]
- Updated dependencies [25f3350]
  - @ai-sdk/provider@0.0.2

## 0.0.2

### Patch Changes

- eb150a6: ai/core: remove scaling of setting values (breaking change). If you were using the temperature, frequency penalty, or presence penalty settings, you need to update the providers and adjust the setting values.
- Updated dependencies [eb150a6]
  - @ai-sdk/provider@0.0.1

## 0.0.1

### Patch Changes

- 7b8791d: Rename baseUrl to baseURL. Automatically remove trailing slashes.
