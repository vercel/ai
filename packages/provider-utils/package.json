{
  "name": "@ai-sdk/provider-utils",
  "version": "2.2.0",
  "license": "Apache-2.0",
  "sideEffects": false,
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist/**/*",
    "test/dist/**/*",
    "CHANGELOG.md"
  ],
  "scripts": {
    "build": "tsup",
    "build:watch": "tsup --watch",
    "clean": "rm -rf dist && rm -rf test/dist",
    "lint": "eslint \"./**/*.ts*\"",
    "type-check": "tsc --noEmit",
    "prettier-check": "prettier --check \"./**/*.ts*\"",
    "test": "pnpm test:node && pnpm test:edge",
    "test:edge": "vitest --config vitest.edge.config.js --run",
    "test:node": "vitest --config vitest.node.config.js --run"
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    },
    "./test": {
      "types": "./test/dist/index.d.ts",
      "import": "./test/dist/index.mjs",
      "module": "./test/dist/index.mjs",
      "require": "./test/dist/index.js"
    }
  },
  "dependencies": {
    "@ai-sdk/provider": "1.1.0",
    "eventsource-parser": "^3.0.0",
    "nanoid": "^3.3.8",
    "secure-json-parse": "^2.7.0"
  },
  "devDependencies": {
    "@types/node": "20.17.24",
    "@vercel/ai-tsconfig": "workspace:*",
    "msw": "2.7.0",
    "tsup": "^8",
    "typescript": "5.6.3",
    "zod": "3.23.8"
  },
  "peerDependencies": {
    "zod": "^3.23.8"
  },
  "engines": {
    "node": ">=18"
  },
  "publishConfig": {
    "access": "public"
  },
  "homepage": "https://sdk.vercel.ai/docs",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vercel/ai.git"
  },
  "bugs": {
    "url": "https://github.com/vercel/ai/issues"
  },
  "keywords": [
    "ai"
  ]
}
