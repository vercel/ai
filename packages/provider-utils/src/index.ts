export * from './combine-headers';
export { convertAsyncIteratorToReadableStream } from './convert-async-iterator-to-readable-stream';
export * from './delay';
export * from './extract-response-headers';
export * from './fetch-function';
export { createIdGenerator, generateId } from './generate-id';
export * from './get-error-message';
export * from './get-from-api';
export * from './is-abort-error';
export * from './load-api-key';
export { loadOptionalSetting } from './load-optional-setting';
export { loadSetting } from './load-setting';
export * from './parse-json';
export * from './post-to-api';
export * from './remove-undefined-entries';
export * from './resolve';
export * from './response-handler';
export * from './uint8-utils';
export * from './validate-types';
export * from './validator';
export * from './without-trailing-slash';

export type { IDGenerator } from './generate-id';
export { parseProviderOptions } from './parse-provider-options';
export type { CoreToolCall, ToolCall } from './types/tool-call';
export type { CoreToolResult, ToolResult } from './types/tool-result';
