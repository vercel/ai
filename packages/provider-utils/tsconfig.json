{
  "extends": "./node_modules/@vercel/ai-tsconfig/ts-library.json",
  "compilerOptions": {
    "composite": true,
    "rootDir": "src",
    "outDir": "dist"
  },
  "exclude": [
    "dist",
    "build",
    "node_modules",
    "tsup.config.ts",
    "test.d.ts"
  ],
  "references": [
    {
      "path": "../provider"
    }
  ]
}
