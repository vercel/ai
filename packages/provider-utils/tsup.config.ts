import { defineConfig } from 'tsup';

export default defineConfig([
  {
    entry: ['src/index.ts'],
    format: ['cjs', 'esm'],
    dts: true,
    sourcemap: true,
    // Keep library target conservative for wide compatibility
    target: 'es2018',
    platform: 'node',
    define: {
      __PACKAGE_VERSION__: JSON.stringify(
        (await import('./package.json', { with: { type: 'json' } })).default
          .version,
      ),
    },
  },
  {
    entry: ['src/test/index.ts'],
    outDir: 'dist/test',
    format: ['cjs', 'esm'],
    dts: true,
    sourcemap: true,
    // Chai uses BigInt literals; ensure the target supports it and avoid bundling chai
    target: 'es2020',
    platform: 'node',
    external: [
      'chai',
      'vitest',
      'vitest/*',
      'msw',
      'msw/*',
      '@vitest/*',
      'vitest/dist/*',
      'vitest/dist/chunks/*',
      'vitest/dist/node/*',
      'vitest/dist/node/chunks/*',
    ],
    define: {
      __PACKAGE_VERSION__: JSON.stringify(
        (await import('./package.json', { with: { type: 'json' } })).default
          .version,
      ),
    },
  },
]);
