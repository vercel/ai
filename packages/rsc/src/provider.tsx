// This file provides the AI context to all AI Actions via AsyncLocalStorage.

import * as React from 'react';
import { InternalAIProvider } from './rsc-shared.mjs';
import {
  withAIState,
  getAIStateDeltaPromise,
  sealMutableAIState,
} from './ai-state';
import type {
  ServerWrappedActions,
  AIAction,
  AIActions,
  AIProvider,
  InternalAIStateStorageOptions,
  OnSetAIState,
  OnGetUIState,
} from './types';

async function innerAction<T>(
  {
    action,
    options,
  }: { action: AIAction; options: InternalAIStateStorageOptions },
  state: T,
  ...args: unknown[]
) {
  'use server';
  return await withAIState(
    {
      state,
      options,
    },
    async () => {
      const result = await action(...args);
      sealMutableAIState();
      return [getAIStateDeltaPromise() as Promise<T>, result];
    },
  );
}

function wrapAction<T = unknown>(
  action: AIAction,
  options: InternalAIStateStorageOptions,
) {
  return innerAction.bind(null, { action, options }) as AIAction<T>;
}

export function createAI<
  AIState = any,
  UIState = any,
  Actions extends AIActions = {},
>({
  actions,
  initialAIState,
  initialUIState,

  onSetAIState,
  onGetUIState,
}: {
  actions: Actions;
  initialAIState?: AIState;
  initialUIState?: UIState;

  /**
   * This function is called whenever the AI state is updated by an Action.
   * You can use this to persist the AI state to a database, or to send it to a
   * logging service.
   */
  onSetAIState?: OnSetAIState<AIState>;

  /**
   * This function is used to retrieve the UI state based on the AI state.
   * For example, to render the initial UI state based on a given AI state, or
   * to sync the UI state when the application is already loaded.
   *
   * If returning `undefined`, the client side UI state will not be updated.
   *
   * This function must be annotated with the `"use server"` directive.
   *
   * @example
   * ```tsx
   * onGetUIState: async () => {
   *   'use server';
   *
   *   const currentAIState = getAIState();
   *   const externalAIState = await loadAIStateFromDatabase();
   *
   *   if (currentAIState === externalAIState) return undefined;
   *
   *   // Update current AI state and return the new UI state
   *   const state = getMutableAIState()
   *   state.done(externalAIState)
   *
   *   return <div>...</div>;
   * }
   * ```
   */
  onGetUIState?: OnGetUIState<UIState>;
}) {
  // Wrap all actions with our HoC.
  const wrappedActions: ServerWrappedActions = {};
  for (const name in actions) {
    wrappedActions[name] = wrapAction(actions[name], {
      onSetAIState,
    });
  }

  const wrappedSyncUIState = onGetUIState
    ? wrapAction(onGetUIState, {})
    : undefined;

  const AI: AIProvider<AIState, UIState, Actions> = async props => {
    if ('useState' in React) {
      // This file must be running on the React Server layer.
      // Ideally we should be using `import "server-only"` here but we can have a
      // more customized error message with this implementation.
      throw new Error(
        'This component can only be used inside Server Components.',
      );
    }

    let uiState = props.initialUIState ?? initialUIState;
    let aiState = props.initialAIState ?? initialAIState;
    let aiStateDelta = undefined;

    if (wrappedSyncUIState) {
      const [newAIStateDelta, newUIState] = await wrappedSyncUIState(aiState);
      if (newUIState !== undefined) {
        aiStateDelta = newAIStateDelta;
        uiState = newUIState;
      }
    }

    return (
      <InternalAIProvider
        wrappedActions={wrappedActions}
        wrappedSyncUIState={wrappedSyncUIState}
        initialUIState={uiState}
        initialAIState={aiState}
        initialAIStatePatch={aiStateDelta}
      >
        {props.children}
      </InternalAIProvider>
    );
  };

  return AI;
}
