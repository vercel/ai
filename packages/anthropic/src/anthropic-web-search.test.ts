import { describe, it, expect } from 'vitest';
import { createTestServer } from '@ai-sdk/provider-utils/test';
import { APICallError } from '@ai-sdk/provider';
import { createAnthropic } from './anthropic-provider';

const TEST_PROMPT = [
  {
    role: 'user' as const,
    content: [{ type: 'text' as const, text: 'What is the latest news?' }],
  },
];

describe('Anthropic Web Search', () => {
  const server = createTestServer({
    'https://api.anthropic.com/v1/messages': {},
  });

  const provider = createAnthropic({
    apiKey: 'test-api-key',
  });
  const model = provider('claude-3-5-sonnet-latest');

  function prepareJsonResponse(body: any) {
    server.urls['https://api.anthropic.com/v1/messages'].response = {
      type: 'json-value',
      body,
    };
  }

  it('should add web search tool when webSearch provider option is provided', async () => {
    prepareJsonResponse({
      type: 'message',
      id: 'msg_test',
      content: [{ type: 'text', text: 'Here are the latest news articles.' }],
      stop_reason: 'end_turn',
      usage: { input_tokens: 10, output_tokens: 20 },
    });

    await model.doGenerate({
      prompt: TEST_PROMPT,
      providerOptions: {
        anthropic: {
          webSearch: {
            maxUses: 3,
            allowedDomains: ['news.com', 'bbc.com'],
          },
        },
      },
    });

    const requestBody = await server.calls[0].requestBodyJson;
    expect(requestBody.tools).toHaveLength(1);
    expect(requestBody.tools[0]).toEqual({
      type: 'web_search_20250305',
      name: 'web_search',
      max_uses: 3,
      allowed_domains: ['news.com', 'bbc.com'],
    });
  });

  it('should add web search tool with user location', async () => {
    prepareJsonResponse({
      type: 'message',
      id: 'msg_test',
      content: [{ type: 'text', text: 'Here are local news.' }],
      stop_reason: 'end_turn',
      usage: { input_tokens: 10, output_tokens: 20 },
    });

    await model.doGenerate({
      prompt: TEST_PROMPT,
      providerOptions: {
        anthropic: {
          webSearch: {
            userLocation: {
              type: 'approximate',
              city: 'San Francisco',
              region: 'California',
              country: 'US',
              timezone: 'America/Los_Angeles',
            },
          },
        },
      },
    });

    const requestBody = await server.calls[0].requestBodyJson;
    expect(requestBody.tools[0].user_location).toEqual({
      type: 'approximate',
      city: 'San Francisco',
      region: 'California',
      country: 'US',
      timezone: 'America/Los_Angeles',
    });
  });

  it('should handle web search results with citations', async () => {
    prepareJsonResponse({
      type: 'message',
      id: 'msg_test',
      content: [
        {
          type: 'server_tool_use',
          id: 'tool_1',
          name: 'web_search',
          input: { query: 'latest AI news' },
        },
        {
          type: 'web_search_tool_result',
          tool_use_id: 'tool_1',
          content: [
            {
              type: 'web_search_result',
              url: 'https://example.com/ai-news',
              title: 'Latest AI Developments',
              encrypted_content: 'encrypted_content_123',
              page_age: 'January 15, 2025',
            },
          ],
        },
        {
          type: 'text',
          text: 'Based on recent articles, AI continues to advance rapidly.',
        },
      ],
      stop_reason: 'end_turn',
      usage: {
        input_tokens: 10,
        output_tokens: 20,
        server_tool_use: { web_search_requests: 1 },
      },
    });

    const result = await model.doGenerate({
      prompt: TEST_PROMPT,
      providerOptions: {
        anthropic: {
          webSearch: { maxUses: 5 },
        },
      },
    });

    expect(result.content).toHaveLength(2);

    expect(result.content[0]).toEqual({
      type: 'source',
      sourceType: 'url',
      id: expect.any(String),
      url: 'https://example.com/ai-news',
      title: 'Latest AI Developments',
      providerMetadata: {
        anthropic: {
          encryptedContent: 'encrypted_content_123',
          pageAge: 'January 15, 2025',
        },
      },
    });

    expect(result.content[1]).toEqual({
      type: 'text',
      text: 'Based on recent articles, AI continues to advance rapidly.',
    });
  });

  it('should handle web search errors', async () => {
    prepareJsonResponse({
      type: 'message',
      id: 'msg_test',
      content: [
        {
          type: 'web_search_tool_result',
          tool_use_id: 'tool_1',
          content: {
            type: 'web_search_tool_result_error',
            error_code: 'max_uses_exceeded',
          },
        },
        {
          type: 'text',
          text: 'I cannot search further due to limits.',
        },
      ],
      stop_reason: 'end_turn',
      usage: { input_tokens: 10, output_tokens: 20 },
    });

    // web search errors should throw an exception in non-streaming mode
    await expect(() =>
      model.doGenerate({
        prompt: TEST_PROMPT,
        providerOptions: {
          anthropic: {
            webSearch: { maxUses: 1 },
          },
        },
      }),
    ).rejects.toThrow(APICallError);
  });

  it('should combine web search with regular tools', async () => {
    prepareJsonResponse({
      type: 'message',
      id: 'msg_test',
      content: [{ type: 'text', text: 'I can search and use tools.' }],
      stop_reason: 'end_turn',
      usage: { input_tokens: 10, output_tokens: 20 },
    });

    await model.doGenerate({
      prompt: TEST_PROMPT,
      tools: [
        {
          type: 'function',
          name: 'calculator',
          description: 'Calculate math',
          parameters: { type: 'object', properties: {} },
        },
      ],
      providerOptions: {
        anthropic: {
          webSearch: { maxUses: 2 },
        },
      },
    });

    const requestBody = await server.calls[0].requestBodyJson;
    expect(requestBody.tools).toHaveLength(2);

    expect(requestBody.tools[1]).toEqual({
      type: 'web_search_20250305',
      name: 'web_search',
      max_uses: 2,
    });

    expect(requestBody.tools[0]).toEqual({
      name: 'calculator',
      description: 'Calculate math',
      input_schema: { type: 'object', properties: {} },
    });
  });
});
