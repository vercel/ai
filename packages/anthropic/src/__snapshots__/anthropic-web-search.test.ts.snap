// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Anthropic Web Search Server-Side Tool > should handle server-side web search results with citations and tool results 1`] = `
[
  {
    "id": "4vrvhg79BuLY9JQU",
    "providerMetadata": {
      "anthropic": {
        "encryptedContent": "encrypted_content_123",
        "pageAge": "January 15, 2025",
      },
    },
    "sourceType": "url",
    "title": "Latest AI Developments",
    "type": "source",
    "url": "https://example.com/ai-news",
  },
  {
    "text": "Based on recent articles, AI continues to advance rapidly.",
    "type": "text",
  },
]
`;
